### ISSUE TITLE
Server returned HTTP response code: 401 for URL:

### ISSUE THREAD

**Issue by AlessandroGnoatto** [sstrickx/yahoofinance-api#130]
Exception in thread "main" java.io.IOException: Server returned HTTP response code: 401 for URL: https://query1.finance.yahoo.com/v7/finance/download/GOOG?period1=1369346400&period2=1527112800&interval=1d&crumb=
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1889)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1884)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1883)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1456)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1440)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
	at yahoofinance.histquotes2.HistQuotes2Request.getResult(HistQuotes2Request.java:132)
	at yahoofinance.Stock.getHistory(Stock.java:327)
	at test.MainClass.main(MainClass.java:47)

**Comment by AlessandroGnoatto** [comment]
Here are the error messages from the logger:

[main] INFO yahoofinance.quotes.query1v7.QuotesRequest - Sending request: https://query1.finance.yahoo.com/v7/finance/quote?symbols=GOOG
[main] WARN yahoofinance.histquotes2.CrumbManager - Failed to set cookie from http request. Historical quote requests will most likely fail.
[main] WARN yahoofinance.histquotes2.CrumbManager - Failed to set crumb from http request. Historical quote requests will most likely fail.
[main] INFO yahoofinance.histquotes2.HistQuotes2Request - Sending request: https://query1.finance.yahoo.com/v7/finance/download/GOOG?period1=1369346400&period2=1527112800&interval=1d&crumb=
[main] WARN yahoofinance.histquotes2.CrumbManager - Failed to set cookie from http request. Historical quote requests will most likely fail.
[main] WARN yahoofinance.histquotes2.CrumbManager - Failed to set crumb from http request. Historical quote requests will most likely fail.

**Comment by sstrickx** [comment]
Hi, the issue is most likely caused by the Oath page they've put in front of their pages in order to request consent to set cookies on your device to use your search, location and browsing data...

Couldn't immediately find the best way to get around this. If anyone has any good ideas, please let me know.

**Comment by giselareichmuth** [comment]
Hi, it seems that when you try from a server outside of the EU, everything still works (not a solution, but a workaround using a vpn server in CH)

**Comment by sstrickx** [comment]
Yea, it looks like they put a consent page in front for the EU due to GDPR... Haven't found a good way yet to get rid of it. Using a VPN outside the EU would indeed be a valid workaround.

**Comment by DanielMartensson** [comment]
I got the same problem.

I used this code:

```
Calendar from = Calendar.getInstance();
            Calendar to = Calendar.getInstance();
            from.add(Calendar.YEAR, -1); // from 1 year ago

            Stock google = YahooFinance.get("GOOG");
            List<HistoricalQuote> googleHistQuotes = google.getHistory(from, to, Interval.DAILY);
            System.out.println("dsd");
            for(int i = 0; i < googleHistQuotes.size(); i++) {
                System.out.println(googleHistQuotes.get(i));
            }
```
Got the 401 error.

**Comment by DanielMartensson** [comment]
Found the solution now. I need to be logged into Yahoo to download historical quotes. So Yahoo Finance API need to have a login option.

**Comment by MariusSiegert** [comment]
I have also tried it with a VPN connection to load historical stock data and it worked. 
Is there a other way to load those data without a VPN? 
@DanielMartensson: Could you give a bit more detailed solution or probably some screenshot or Java code?
Thanks a lot!

**Comment by couclock** [comment]
It works also with a simple anonymous proxy (get one on https://free-proxy-list.net/ for example). I hard set environment variables at the beginning of my main method using : 
```
System.setProperty("http.proxyHost", "35.199.64.211");
System.setProperty("http.proxyPort", "80");
System.setProperty("https.proxyHost", "35.199.64.211");
System.setProperty("https.proxyPort", "80");
```
It's slower than without proxy but there is not 401 error response anymore.

Good work @sstrickx !

**Comment by DanielMartensson** [comment]
@MariusSiegert I got the 401 error and if I go to Yahoo Finance and try to download the historical quotes, I still got the 401 error in my web browser. I need to login to Yahoo and then I can download. 

For this API. I got this error. https://imgur.com/a/wtz9any

By using this code:

```
Calendar from = Calendar.getInstance();
            Calendar to = Calendar.getInstance();
            from.add(Calendar.YEAR, -1); // from 1 year ago

            Stock google = YahooFinance.get("GOOG");
            List<HistoricalQuote> googleHistQuotes = google.getHistory(from, to, Interval.DAILY);
            System.out.println("dsd");
            for(int i = 0; i < googleHistQuotes.size(); i++) {
                System.out.println(googleHistQuotes.get(i));
            }

```

**Comment by matoos32** [comment]
I experienced the same problem but was able to get around it not by signing in to Yahoo or by using a VPN, but by changing the DNS server my computer users. I was using a specific server address and when I reverted to obtaining one automatically I had no more issues.

**Comment by SekoDevl** [comment]
Hi, 

Depends on your country, you have to accept cookie. 
I add these codes at line 65 on the method "setCookie" of the class "CrumbManager": 


```
		//  If cookie is not set, we should accept to activate cookie
        InputStreamReader is = new InputStreamReader(connection.getInputStream());
        BufferedReader br = new BufferedReader(is);
        String line;
        Pattern patternPostForm = Pattern.compile("(.*)(method=\"post\" action=\"/consent\")(.*)");
        Pattern patternInput = Pattern.compile("(.*)(<input type=\"hidden\" name=\")(.*?)(\" value=\")(.*?)(\">)");
        Matcher matcher;
        Map<String,String> datas = new HashMap<String,String>();
        boolean postFind = false;
        // Read source to get params data for post request
        while( (line =br.readLine())!=null ) {
        	matcher = patternPostForm.matcher(line);
        	if(matcher.find()){
        		postFind = true;
        	}
        	
        	if(postFind){
        		matcher = patternInput.matcher(line);
        		if(matcher.find()){
        			String name = matcher.group(3);
        			String value = matcher.group(5);        			
        			datas.put(name, value);		
        		}
        	}        	
           
        }
        // If params are not empty, send the post request
        if(datas.size()>0){
        	 log.debug("POST https://guce.oath.com/consent");
        	 datas.put("namespace","yahoo");
        	 datas.put("agree","agree");
        	 datas.put("originalDoneUrl",YahooFinance.HISTQUOTES2_SCRAPE_URL);
        	 datas.put("doneUrl","https://guce.yahoo.com/copyConsent?sessionId="+datas.get("sessionId")+"&inline="+datas.get("inline")+"&lang="+datas.get("locale"));
			 log.debug(connection.getURL().toString());
			
        	 URL requestOath = new URL("https://guce.oath.com/consent");
        	 HttpURLConnection connectionOath = null;
        	 connectionOath = (HttpURLConnection) requestOath.openConnection();
        	 connectionOath.setConnectTimeout(YahooFinance.CONNECTION_TIMEOUT);
        	 connectionOath.setReadTimeout(YahooFinance.CONNECTION_TIMEOUT);
        	 connectionOath.setRequestMethod( "POST" );
        	 connectionOath.setDoOutput( true );
        	 connectionOath.setRequestProperty("Referer", connection.getURL().toString());
        	 connectionOath.setRequestProperty("Host","guce.oath.com");
        	 connectionOath.setRequestProperty("Origin","https://guce.oath.com");
        	 connectionOath.setRequestProperty("Content-Type","application/x-www-form-urlencoded");
        	 StringBuilder params=new StringBuilder("");
        	
    		 for ( String key : datas.keySet() ) {
    			 if(params.length() == 0 ){
        			 params.append(key);
        			 params.append("=");
        			 params.append(URLEncoder.encode(datas.get(key),"UTF-8"));
    			 }else{
        			 params.append("&");
        			 params.append(key);
        			 params.append("=");
        			 params.append(URLEncoder.encode(datas.get(key),"UTF-8"));
    				 
    			 }
    		  }
        		 
        	      
        	 log.debug("Params = "+ params.toString());      
        	 connectionOath.setRequestProperty("Content-Length",Integer.toString(params.toString().length()));
        	 OutputStreamWriter outputStreamWriter = new OutputStreamWriter(connectionOath.getOutputStream());
             outputStreamWriter.write(params.toString());
             outputStreamWriter.flush();
             connectionOath.setInstanceFollowRedirects(true);
             connectionOath.getResponseCode();
        }
        
        // Then Set the cookie with the cookieJar
        CookieStore cookieJar =  ((CookieManager)CookieHandler.getDefault()).getCookieStore();
        List <HttpCookie> cookies = cookieJar.getCookies();
        for (HttpCookie hcookie: cookies) {        	
        	if(hcookie.toString().matches("B=.*")) {
                 cookie = hcookie.toString();
                 log.debug("Set cookie from http request: {}", cookie);
                 return;
             }
        }


```
``

**Comment by SekoDevl** [comment]
Done

**Comment by anarayn** [comment]
When will this fix be released?

**Comment by sstrickx** [comment]
The issue should be fixed now in version 3.14.0.
Thanks a lot to @SekoDevl for the pull request!
